state/core_access.go

```diff
@@ -174,24 +192,74 @@ func (ca *CoreAccessor) SubmitPayForBlob(
 		appblobs[i] = &b.Blob
 	}
 
-	response, err := appblob.SubmitPayForBlob(
-		ctx,
-		ca.signer,
-		ca.coreConn,
-		appblobs,
-		apptypes.SetGasLimit(gasLim),
-		withFee(fee),
-	)
-	// metrics should only be counted on a successful PFD tx
-	if err == nil && response.Code == 0 {
-		ca.lastPayForBlob = time.Now().UnixMilli()
-		ca.payForBlobCount++
+	// if the min gas price is not set we query the node that we are connected for it
+	if ca.minGasPrice < 0 {
+		minGasPrice, err := ca.queryMinimumGasPrice(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("querying minimum gas price: %w", err)
+		}
+		ca.minGasPrice = minGasPrice
+	}
+
+	// we only estimate gas if the user wants us to (by setting the gasLim to 0). In the future we may want
+	// to make these arguments optional.
+	if gasLim == 0 {
+		blobSizes := make([]uint32, len(blobs))
+		for i, blob := range blobs {
+			blobSizes[i] = uint32(len(blob.Data))
+		}
+
+		// TODO: the default gas per byte and the default tx size cost per byte could be changed through governance
+		// This section could be more robust by tracking these values and adjusting the gas limit accordingly
+		// (as is done for the gas price)
+		gasLim = apptypes.EstimateGas(blobSizes, appconsts.DefaultGasPerBlobByte, auth.DefaultTxSizeCostPerByte)
 	}
 
-	if response != nil && response.Code != 0 {
-		err = errors.Join(err, sdkErrors.ABCIError(response.Codespace, response.Code, response.Logs.String()))
+	var (
+		lastErr      error
+		estimatedFee = false
+	)
+	// set the fee for the user as the minimum gas price multiplied by the gas limit
+	if fee.IsNegative() {
+		estimatedFee = true
+		fee = sdktypes.NewInt(int64(math.Ceil(ca.minGasPrice * float64(gasLim))))
+	}
+
+	for attempt := 0; attempt < maxRetries; attempt++ {
+		response, err := appblob.SubmitPayForBlob(
```

Comment:
Let's fix https://github.com/celestiaorg/celestia-app/issues/1885, while we are on it.
This also means the comment on SubmitPayForBlob method will need to be slightly modified.